transports/webrtc/src/udp_mux.rs

```diff
@@ -0,0 +1,415 @@
+// MIT License
+//
+// Copyright (c) 2021 WebRTC.rs
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use std::{
+    borrow::Cow,
+    collections::{HashMap, HashSet},
+    io::ErrorKind,
+    net::SocketAddr,
+    sync::Arc,
+};
+
+use futures::channel::mpsc;
+use libp2p_core::multiaddr::{Multiaddr, Protocol};
+use webrtc_ice::udp_mux::UDPMux;
+use webrtc_util::{sync::RwLock, Conn, Error};
+
+use tokio_crate as tokio;
+use tokio_crate::sync::{watch, Mutex};
+
+mod socket_addr_ext;
+
+mod udp_mux_conn;
+use udp_mux_conn::{UDPMuxConn, UDPMuxConnParams};
+
+use async_trait::async_trait;
+
+use stun::{
+    attributes::ATTR_USERNAME,
+    message::{is_message as is_stun_message, Message as STUNMessage},
+};
+
+const RECEIVE_MTU: usize = 8192;
+
+/// Normalize a target socket addr for sending over a given local socket addr. This is useful when
+/// a dual stack socket is used, in which case an IPv4 target needs to be mapped to an IPv6
+/// address.
+fn normalize_socket_addr(target: &SocketAddr, socket_addr: &SocketAddr) -> SocketAddr {
+    match (target, socket_addr) {
+        (SocketAddr::V4(target_ipv4), SocketAddr::V6(_)) => {
+            let ipv6_mapped = target_ipv4.ip().to_ipv6_mapped();
+
+            SocketAddr::new(std::net::IpAddr::V6(ipv6_mapped), target_ipv4.port())
+        },
+        // This will fail later if target is IPv6 and socket is IPv4, we ignore it here
+        (_, _) => *target,
+    }
+}
```

Comment:
I'd prefer utility functions to be at the bottom of files due to the lower importance.